### PR TITLE
[FIX] mrp: improve finding bom_kits in product _compute_quantities_dict

### PR DESCRIPTION
**Pull request by Aurelienvd**
Replace calling _bom_find for each product in self by first searching
on mrp.bom on every product_id or product_tmpl_id in self. This is done to avoid doing 
one select query per product in self with limit=1.

#### Speed-up
The expected speed-up on a client DB with 11000 product.product and 8000 boms
is from 40s to 4s for the `Negative Forecasted Quantities` filter in stock kanban/list views.

Small discussion about a testing edge case below in https://github.com/odoo/odoo/pull/68210#issuecomment-804177217

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by Aurelienvd** [comment]
#### Testing

Apart from testing the nfq filter, I also tested this PR in a new database. The idea was to see if the `bom_kits` variable
was the same before and after applying the changes. While it is indeed the same when there are products with variants and
bom (with or without different sequence values) for either the product.templates or the product.product, there is an "edge case"
that I think I should report here.

Suppose that there are 3 products `product.template(1,2,3)`, each with two variants `product.product(1,2,3,4,5,6)`. 
Suppose also that there are 2 bom for each template and 2 bom for each variant `mrp.bom(1,2,3,...,18)` and that each bom has the same sequence value `sequence = 0`. Searching for negative forecasted quantities, the size of `bom_kits` should be 6, one for each `product.product`. 

Now, for `product.product(1)`, there will be 2 boms `mrp.bom(1,2)` for the template and 2 boms for the product itself `mrp.bom(3,4)`. According to the search criteria of `mrp_bom._bom_find`, both bom 3 and 4 are equivalent because they both
have the same sequence, product_id and product_tmpl_id. In this case, there is now guarantee that the bom returned by `_get_bom_kits` will be the same as the bom returned by `_bom_find`. The latter could return `mrp.bom(3)` for  `product.product(1)` while the former could return `mrp.bom(3)`.

I don't think this is an issue because here the purpose is just to return one bom per product.product respecting some searching
criteria. So in case of ties, the tying boms are interchangeable. But I may be wrong about that, hence the reporting of this "edge case" here.

**Comment by yelizariev** [comment]
I made the same for v13 :laughing: 
https://github.com/odoo/odoo/pull/68229

File: addons/mrp/models/product.py [hunk 1]
```diff
@@ -113,6 +113,33 @@ def _compute_mrp_product_qty(self):
                 continue
             product.mrp_product_qty = float_round(mapped_data.get(product.id, 0), precision_rounding=product.uom_id.rounding)
 
+    def _get_bom_kits(self):
+        products = self.filtered(lambda p: p.type != 'service' or p.product_tmpl_id.type != 'service')
```
Comment (yelizariev) [inline addons/mrp/models/product.py:117]: ``p.type`` and ``p.product_tmpl_id.type`` are always the same, no?

File: addons/mrp/models/product.py [hunk 2]
```diff
@@ -113,6 +113,33 @@ def _compute_mrp_product_qty(self):
...
+        if not products:
+            return {}
+
+        def get_bom(kits, product_id, produt_tmpl_id):
+            template_bom = kits.get((False, produt_tmpl_id), {'sequence': float('inf')})
+            product_bom = kits.get((product_id, produt_tmpl_id), {'sequence': float('inf')})
+            return template_bom.get('bom') if template_bom['sequence'] < product_bom['sequence'] else product_bom.get('bom')
```
Comment (yelizariev) [inline addons/mrp/models/product.py:124]: I think I forgot to check sorting in my own PR
https://github.com/odoo/odoo/pull/68229#pullrequestreview-618484779

File: addons/mrp/models/product.py [hunk 3]
```diff
@@ -113,6 +113,33 @@ def _compute_mrp_product_qty(self):
...
+        product_domain = ['|',
+            ('product_id', 'in', products.ids),
+            '&', ('product_id', '=', False), ('product_tmpl_id', 'in', products.product_tmpl_id.ids)]
+        kits = {}
+        for bom in self.env['mrp.bom'].search(product_domain + company_domain + [('type', '=', 'phantom')], order='sequence, product_id'):
+            kits.setdefault((bom.product_id.id, bom.product_tmpl_id.id), {"bom": bom, "sequence": bom.sequence})
+        return {
+            product: bom
+            for product in products
+            for bom in (get_bom(kits, product.id, product.product_tmpl_id.id),)
```
Comment (yelizariev) [inline addons/mrp/models/product.py:139]: Interesting way to make assignment in such for blocks

**Pull request closed without merging**